File: P/64_Minimum_Path_Sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        for(int i = 0; i < grid.length; i++) {
            if(i + 1 < grid.length) {
                grid[i + 1][0] += grid[i][0];
            }
            for(int j = 0; j < grid[0].length - 1; j++) {
                if(i == 0) {
                    grid[i][j + 1] += grid[i][j];
                } else {
                    grid[i][j + 1] += grid[i - 1][j + 1] < grid[i][j] ? grid[i - 1][j + 1] : grid[i][j];
                }
            }
        }
        return grid[grid.length - 1][grid[0].length - 1];
    }
}

class Solution {
    public int minPathSum(int[][] grid) {
        for(int i = 0; i < grid[0].length - 1; i++)
            grid[0][i + 1] += grid[0][i];
        
        for(int i = 1; i < grid.length; i++) {
            grid[i][0] += grid[i - 1][0];
            for(int j = 0; j < grid[0].length - 1; j++)
                grid[i][j + 1] += grid[i - 1][j + 1] < grid[i][j] ? grid[i - 1][j + 1] : grid[i][j];
        }
        return grid[grid.length - 1][grid[0].length - 1];
    }
}

File: 주영/DP/64. Minimum Path Sum.js
var minPathSum = function (grid) {
  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[i].length; j++) {
      if (i === 0 && j !== 0) grid[i][j] += grid[i][j - 1];
      if (i !== 0 && j === 0) grid[i][j] += grid[i - 1][j];
      if (i !== 0 && j !== 0)
        grid[i][j] = grid[i][j] + Math.min(grid[i - 1][j], grid[i][j - 1]);
    }
  }

  let len = grid.length;

  console.log(grid[len - 1][len - 1]);
};

let grid = [
  [1, 2, 3],
  [4, 5, 6],
];
minPathSum(grid);
